### PR TITLE
Add support for path-based routing

### PR DESCRIPTION
**Pull request by gregsymons**
This PR adds the ability to serve multiple containers through the same vhost using path-based routes. This can be useful if you have several independent services that you want make appear as one.

The path-based proxies assume that the incoming request uri should be mapped to the root of the container, and should be rewritten to remove the prefix on the proxy host. This PR does not provide support for either passing the full request uri through to the backend nor for requesting any context other than the root of the backend container.

In other words, the incoming path is always passed to `http://<backend-address>:<backend-port>/`.

### CODE REVIEW

File: nginx.tmpl [hunk 1]
```diff
@@ -17,6 +17,55 @@
...
+		include /etc/nginx/vhost.d/default_location;
+	{{ end }}
+}
+{{ end }}
+
+{{ define "upstream-definition" }}
+	upstream {{ .Upstream }} {
+		{{ range $container := .Containers }}
+			{{ $addrLen := len $container.Addresses }}
+			{{ $networks := .Networks }}
```
Comment (bmitch3020) [inline nginx.tmpl]: I believe this needs to be moved up 2 lines. Otherwise $networks is matching all container networks and listing every IP, not just the ones on the same network as the proxy server.

**Comment by narfanator** [comment]
Does this support specifying a `VIRTUAL_PATH` with no `VIRTUAL_HOST`?

**Comment by armpogart** [comment]
What's the progress with this PR?

**Comment by gabadi** [comment]
+1

**Comment by Hadlock** [comment]
+1

File: nginx.tmpl (hunk 1 above)
Comment (davidneudorfer) [inline nginx.tmpl]: @gregsymons can you update this PR?

**Comment by gregsymons** [comment]
Should this be closed in favor of #254?

**Comment by ThejanW** [comment]
What's the progress with this PR?

**Comment by StephenOTT** [comment]
anyone using this in production?

**Comment by jhu-arod** [comment]
much needed feature.  +1


I do have a scenario where I would like a single docker instance to service more than one path.  I don't believe this would cover that.  Is there some way to make this possible?

**Comment by rustyx** [comment]
Looks like @jwilder abandoned us...
Just use this fork instead:
https://hub.docker.com/r/gregsymons/docker-nginx-proxy/

**Comment by janek-moveapp** [comment]
+1

**Comment by scatat** [comment]
Yes please - can we merge it?

**Comment by delfer** [comment]
@thomasleveil @kamermans @jwilder please review and merge if possible. We all really need it!

**Comment by thomasleveil** [comment]
sorry, I'm not contributing to this project anymore. I moved to Traefik

**Comment by delfer** [comment]
Hmm, I think same configuration can be reached using Traefik:

- it can interact with Docker https://docs.traefik.io/user-guides/docker-compose/basic-example/
- can route using path https://docs.traefik.io/routing/routers/#rule
- have embedded Let's Encrypt support https://docs.traefik.io/user-guides/docker-compose/acme-http/
- have official ARM docker image (for Raspberry PI)
I will try to go on it.

**Comment by KunalSolanke** [comment]
I tried the updated tmpl file with my setup but something weird is happening with the end slashes
VIRTUAL_PATH 
Screenshot from 2021-05-27 01-09-50

Note the slashes after patjj
Screenshot from 2021-05-27 01-10-44
Screenshot from 2021-05-27 01-11-01
Screenshot from 2021-05-27 01-11-20

Notice the slashes at the end,I had set VIRTUAL_PATH=/test,when I enter /tests in browser it tries to go to localhost/s ans similarly it add two // at the end for other urls

This is my global proxy setup

Screenshot from 2021-05-27 01-12-04

**Comment by buchdag** [comment]
@KunalSolanke this PR won't be merged, please see #1607 for the current, worked on version.

**Comment by buchdag** [comment]
This features has been added to the `dev` branch with #1607 

It's available now on the `dev` images `nginxproxy/nginx-proxy:dev` and `nginxproxy/nginx-proxy:dev-alpine`. Testing and feedback are welcome, the docs for the feature is [here](https://github.com/nginx-proxy/nginx-proxy/tree/dev#path-based-routing) and there is [another PR](https://github.com/nginx-proxy/nginx-proxy/pull/1699) opened for additional improvements and comments.
